01/ht/HappyCodeHere/learn.javascript.ru/Замыкания, область видимости/Локальные переменные для объекта.md

```diff
@@ -0,0 +1,213 @@
+  	return a + b;
+  }
+}
+```
+
+## Функция - строковый буфер
+
+В некоторых языках программирования существует объект «строковый буфер», который аккумулирует внутри себя значения. Его функционал состоит из двух возможностей:
+
+Добавить значение в буфер.
+Получить текущее содержимое.
+Задача – реализовать строковый буфер на функциях в JavaScript, со следующим синтаксисом:
+
+Создание объекта: var buffer = makeBuffer();.
+Вызов makeBuffer должен возвращать такую функцию buffer, которая при вызове buffer(value) добавляет значение в некоторое внутреннее хранилище, а при вызове без аргументов buffer() – возвращает его.
+
+Вот пример работы:
+
+```javascript
+function makeBuffer() { /* ваш код */ }
+
+var buffer = makeBuffer();
+
+// добавить значения к буферу
+buffer('Замыкания');
+buffer(' Использовать');
+buffer(' Нужно!');
+
+// получить текущее значение
+alert( buffer() ); // Замыкания Использовать Нужно!
+```
+
+Буфер должен преобразовывать все данные к строковому типу:
+
+```javascript
+var buffer = makeBuffer();
+buffer(0);
+buffer(1);
+buffer(0);
+
+alert( buffer() ); // '010'
+```
+
+Решение не должно использовать глобальные переменные.
+
+### Строковый буфер с очисткой
+
+Добавьте буферу из решения задачи Функция - строковый буфер метод buffer.clear(), который будет очищать текущее содержимое буфера:
+
+```javascript
+function makeBuffer() {
+  ...ваш код...
+}
+
+var buffer = makeBuffer();
+
+buffer("Тест");
+buffer(" тебя не съест ");
+alert( buffer() ); // Тест тебя не съест
+
+buffer.clear();
+
+alert( buffer() ); // ""
+```
+
+#### Решение:
+
+```javascript
+function makeBuffer() {
+  var buffer = '';
+
+
+  boo.clear = function() {
+  	buffer = '';
+  }
+
+  function boo(text) {
+  	if (!text) return buffer;
+  	buffer += String(text);
+  }
+
+  return boo;
+
+}
+```
+
+## Сортировка
+
+У нас есть массив объектов:
+
+```javascript
+var users = [{
+  name: "Вася",
+  surname: 'Иванов',
+  age: 20
+}, {
+  name: "Петя",
+  surname: 'Чапаев',
+  age: 25
+}, {
+  name: "Маша",
+  surname: 'Медведева',
+  age: 18
+}];
+```
+
+Обычно сортировка по нужному полю происходит так:
+
+```javascript
+// по полю name (Вася, Маша, Петя)
+users.sort(function(a, b) {
+  return a.name > b.name ? 1 : -1;
+});
+
+// по полю age  (Маша, Вася, Петя)
+users.sort(function(a, b) {
+  return a.age > b.age ? 1 : -1;
+});
+```
+
+Мы хотели бы упростить синтаксис до одной строки, вот так:
+
+```javascript
+users.sort(byField('name'));
+users.forEach(function(user) {
+  alert( user.name );
+}); // Вася, Маша, Петя
+
+users.sort(byField('age'));
+users.forEach(function(user) {
+  alert( user.name );
+}); // Маша, Вася, Петя
+```
+
+То есть, вместо того, чтобы каждый раз писать в sort function... – будем использовать byField(...)
+
+Напишите функцию byField(field), которую можно использовать в sort для сравнения объектов по полю field, чтобы пример выше заработал.
+
+#### Решение:
+
+```javascript
+function byField(field) {
+	return function(a, b) {
```

Comment:
![https://www.dropbox.com/s/lm6ix2x54w8piwv/%D0%A1%D0%BA%D1%80%D0%B8%D0%BD%D1%88%D0%BE%D1%82%202017-05-25%2010.47.05.jpg?dl=1](https://www.dropbox.com/s/lm6ix2x54w8piwv/%D0%A1%D0%BA%D1%80%D0%B8%D0%BD%D1%88%D0%BE%D1%82%202017-05-25%2010.47.05.jpg?dl=1)